feature: implement shared timer functionality for games

Noor-Serry opened the pull request:
This commit introduces a shared timer component and related logic for use in game screens.

Key changes:

- Added `TimerComponent.kt`: A new Composable that displays a circular progress timer with an animated color (green to red) based on the remaining time.
- Added `BaseGameUiState.kt`: Defines the UI state for the timer, including current time, color, and progress.
- Added `BaseGameInteractionListener.kt`: An interface for timer interactions (start, stop).
- Added `BaseGameViewModel.kt`: A base ViewModel that implements `BaseGameInteractionListener` and manages the timer logic using coroutines. It provides callbacks for timer updates and completion.
- Added string resources for "second" (singular and plural) for timer display.
- Updated `GameCard.kt`: Fixed a bug where the `onCardClick` was triggered when `isPlayable` was false.
- Added `androidx.ui.graphics.android` dependency for graphics operations.

# Pull Request Template

## Description
Provide a clear and concise description of what this pull request does. Include the purpose and context for the changes.

---

## Changes Made
List the changes introduced in this pull request:

-

---

## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.

https://github.com/user-attachments/assets/7b378d21-645d-4d9f-8351-b587f76adfc1



---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
this video just for showing the timer I am not embedded it in lets play screen

Noor-Serry committed:
feature: implement shared timer functionality for games

This commit introduces a shared timer component and related logic for use in game screens.

Key changes:

- Added `TimerComponent.kt`: A new Composable that displays a circular progress timer with an animated color (green to red) based on the remaining time.
- Added `BaseGameUiState.kt`: Defines the UI state for the timer, including current time, color, and progress.
- Added `BaseGameInteractionListener.kt`: An interface for timer interactions (start, stop).
- Added `BaseGameViewModel.kt`: A base ViewModel that implements `BaseGameInteractionListener` and manages the timer logic using coroutines. It provides callbacks for timer updates and completion.
- Added string resources for "second" (singular and plural) for timer display.
- Updated `GameCard.kt`: Fixed a bug where the `onCardClick` was triggered when `isPlayable` was false.
- Added `androidx.ui.graphics.android` dependency for graphics operations.

ahmedmahgoub0 commented on ui/src/main/java/com/amsterdam/ui/components/GameCard.kt:
```diff
@@ -82,7 +82,7 @@ fun GameCard(
                     brush = Brush.horizontalGradient(colors = borderColors),
                     shape = RoundedCornerShape(16.dp),
-                ).clickable { if (!isPlayable) onCardClick() },
+                ).clickable { if (isPlayable) onCardClick() },
```
you can pass `enabled` boolean for `clickable` as clickable(enabled = isPlayable) { ... }
because `clickable { if (isPlayable) onCardClick() }` means that it gives clicking effect but do nothing

Noor-Serry replied:
great

ahmedmahgoub0 commented on ui/src/main/java/com/amsterdam/ui/components/guessGame/TimerComponent.kt:
```diff
@@ -0,0 +1,63 @@
+    baseGameUiState: BaseGameUiState,
+    modifier: Modifier = Modifier
+) {
+    val targetColor = when (baseGameUiState.currentTimerColor) {
```
pass when statement directly to animation to avoid recompositions

Noor-Serry replied:
I do not think that will avoid the recompositing but still I can improve it  with 
remember + derivedStateOf

ahmedmahgoub0 replied:
no need for a variable, but ok

Noor-Serry replied:
I think if we  put the when inside the animation it will recalculated with each recomposetion for TimerComponent that like what was I did but now with remember and drived state of 
it just will calculate if the currentTimerColor changed

Noor-Serry committed:
refactor: optimize TimerComponent color updates and GameCard clickability

This commit introduces two performance optimizations:

- In `TimerComponent.kt`, `derivedStateOf` is now used with `remember` for `targetColor`. This ensures that the color calculation only happens when `baseGameUiState.currentTimerColor` actually changes, preventing unnecessary recompositions.
- In `GameCard.kt`, the `clickable` modifier now uses the `enabled` parameter, set to `isPlayable`. This avoids unnecessary recompositions when `onCardClick` changes if the card is not playable.

Additionally, the timer progress calculation in `BaseGameViewModel.kt` has been corrected to use `remainingSeconds` instead of `remainingMillis` for a more accurate progress representation.

eslam-99 commented on ui/src/main/java/com/amsterdam/ui/components/GameCard.kt:
This GameCard should be in guessGame components not the ui components, right?

Noor-Serry replied:
that is true

eslam-99 replied:
It still there in old place😂

Noor-Serry replied:
no its moved 
<img width="395" height="382" alt="image" src="https://github.com/user-attachments/assets/d72aeb4d-b672-4c02-9c80-55b848d8cea8" />

saraashraf11 commented on ui/src/main/java/com/amsterdam/ui/components/guessGame/TimerComponent.kt:
```diff
@@ -0,0 +1,71 @@
+                .zIndex(1f)
+        )
+
+        Canvas(modifier = Modifier.size(48.dp).zIndex(10f)) {
```
Why is the canvas size larger than the box size?

Noor-Serry replied:
I think I forget to change it

saraashraf11 commented on viewModel/src/main/java/com/amsterdam/viewmodel/sharedGame/BaseGameInteractionListiner.kt:
```diff
@@ -0,0 +1,11 @@
+
+interface BaseGameInteractionListener {
+    fun startTheTimer(
+        questionSecondsLimit: Int,
```
I think it would be clearer to replace the name questionSecondsLimit with questionTimeLimitSeconds. What do you think?

Noor-Serry replied:
okay good name

saraashraf11 commented on viewModel/src/main/java/com/amsterdam/viewmodel/sharedGame/BaseGameViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+import kotlinx.coroutines.launch
+import kotlin.math.max
+
+open class BaseGameViewModel<S, E>(
```
Can you explain what you mean by <S, E>?

Noor-Serry replied:
S for State and E for Effect

saraashraf11 replied:
Can you write them as state and effect?

Noor-Serry replied:
no most generic type written as character also we have the same thing in the base view model

Noor-Serry committed:
refactor: improve timer logic and UI components

This commit introduces several refinements:

- Renamed `questionSecondsLimit` to `questionTimeLimitSeconds` in `BaseGameInteractionListener.kt` and `BaseGameViewModel.kt` for clarity.
- Adjusted the `TimerComponent.kt`:
    - Changed `Canvas` size from `48.dp` to `40.dp`.
    - Changed `strokeWidth` from `5.dp` to `3.dp`.
- Moved `GuessCard.kt` from `com.amsterdam.ui.components.guessGame` to `com.amsterdam.ui.screens.letsPlay.component`.
- Updated imports in `GuessPicture.kt` and `GuessTitle.kt` to reflect the new location of `GuessCard`.

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/sharedGame/BaseGameUiState.kt:
```diff
@@ -0,0 +1,11 @@
+    val currentTimerColor: TimerColor = TimerColor.RED,
+    val progress: Float = 0f
+) {
+    enum class TimerColor() {
```
The ui state in view model shouldn't know about the colors that will be  displayed in the screens.

The UI can handle this based on the data in UI state I think

Noor-Serry replied:
I didn't consider that the UI is designed to be dumb, so it shouldn't contain any if conditions or be aware of thresholds.

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/sharedGame/BaseGameViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+    private var endTimeMillis: Long = 0L
+    private var isStopTimerClicked: Boolean = false
+
+    override fun startTheTimer(
```
Not responsibility for the view model to handle the implementation of timer

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/sharedGame/BaseGameViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+import kotlinx.coroutines.launch
+import kotlin.math.max
+
+open class BaseGameViewModel<S, E>(
```
I think no need for this view model as it is not making anything related to the view model It will be a handler or manager for timer

Noor-Serry committed:
refactor: move timer logic to domain layer and rename UI state

This commit refactors the timer functionality by:

- Moving the timer logic from `BaseGameViewModel` to a new `TimerHandler` class in the domain layer. This promotes better separation of concerns and reusability.
- Renaming `BaseGameUiState` to `TimerUiState` for clarity.
- Deleting the now obsolete `BaseGameViewModel` and `BaseGameInteractionListener`.
- Updating `TimerComponent.kt` to use `TimerUiState`.

The `TimerHandler` class encapsulates the timer start, stop, and update logic, taking a `CoroutineScope` as a dependency. It provides callbacks for timer updates and completion.

Noor-Serry committed:
refactor: move TimerHandler to domain.timer package

The `TimerHandler` class has been moved from the `com.amsterdam.domain.useCase.timer` package to `com.amsterdam.domain.timer`.

Additionally, the `TimerHandler` class is no longer `open`.
